### ISSUE TITLE
Object of type "None" cannot be used as iterable value

### ISSUE THREAD

**Issue by vincentmasse** [microsoft/pylance-release#3607]
## Environment data

-   Language Server version: v2022.11.20
-   OS and version: Ubuntu 22
-   Python version (& distribution if applicable, e.g. Anaconda): venv 3.8.10

## Code Snippet

```python
def validation_not_none(field: Optional[Any]) -> bool:
    return field is not None


def test(value: Optional[List[bool]]):
    if validation_not_none(value):
        if any(v for v in value):
            print("ok")
```

## Expected behavior

It should not have any warning

## Actual behavior

warning about "Object of type "None" cannot be used as iterable value"
In the code value isn't None if it enters in the first `if`

## Logs

```
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: setFileOpened
(18524) Background analysis message: markFilesDirty
(18524) [FG] parsing: /home/user/file.py (12ms)
(18524) [FG] binding: /home/user/file.py (6ms)
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: analyze
(18524) [BG(1)] analyzing: /home/user/file.py ...
(18524) [BG(1)]   parsing: /home/user/file.py (15ms)
(18524) [BG(1)]   binding: /home/user/file.py (5ms)
(18524) [BG(1)]   checking: /home/user/file.py (89ms)
(18524) [BG(1)] analyzing: /home/user/file.py (110ms)
(18524) Background analysis message: getSemanticTokens delta
(18524) [BG(1)] getSemanticTokens delta previousResultId:1668106580187 at /home/user/file.py (14ms)
(18524) Background analysis message: resumeAnalysis
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: setFileOpened
(18524) Background analysis message: markFilesDirty
(18524) [FG] completion at /home/user/file.py:12:27 ...
(18524) [FG]   parsing: /home/user/file.py (12ms)
(18524) [FG]   binding: /home/user/file.py (1ms)
(18524) [FG] completion at /home/user/file.py:12:27 [found 11 items] (25ms)
(18524) Background analysis message: setFileOpened
(18524) Background analysis message: markFilesDirty
(18524) [FG] completion at /home/user/file.py:12:28 ...
(18524) [FG]   parsing: /home/user/file.py (11ms)
(18524) [FG]   binding: /home/user/file.py (4ms)
(18524) [FG] completion at /home/user/file.py:12:28 [found 5 items] (25ms)
(18524) Background analysis message: setFileOpened
(18524) [FG] completion at /home/user/file.py:12:29 ...
(18524) [FG]   parsing: /home/user/file.py (8ms)
(18524) [FG]   binding: /home/user/file.py (1ms)
(18524) [FG] completion at /home/user/file.py:12:29 [found 4 items] (22ms)
(18524) Background analysis message: markFilesDirty
(18524) Background analysis message: setFileOpened
(18524) Background analysis message: markFilesDirty
(18524) [FG] parsing: /home/user/file.py (6ms)
(18524) [FG] binding: /home/user/file.py (1ms)
(18524) Background analysis message: getSemanticTokens delta
(18524) [BG(1)] getSemanticTokens delta previousResultId:1668106621507 at /home/user/file.py ...
(18524) [BG(1)]   parsing: /home/user/file.py (9ms)
(18524) [BG(1)]   binding: /home/user/file.py (2ms)
(18524) [BG(1)] getSemanticTokens delta previousResultId:1668106621507 at /home/user/file.py (85ms)
(18524) Background analysis message: setFileOpened
(18524) Background analysis message: markFilesDirty
(18524) [FG] parsing: /home/user/file.py (4ms)
(18524) [FG] binding: /home/user/file.py (2ms)
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: setFileOpened
(18524) Background analysis message: markFilesDirty
(18524) Background analysis message: getSemanticTokens delta
(18524) [BG(1)] getSemanticTokens delta previousResultId:1668106623389 at /home/user/file.py ...
(18524) [BG(1)]   parsing: /home/user/file.py (3ms)
(18524) [BG(1)]   binding: /home/user/file.py (2ms)
(18524) [FG] parsing: /home/user/file.py (5ms)
(18524) [FG] binding: /home/user/file.py (2ms)
(18524) [BG(1)] getSemanticTokens delta previousResultId:1668106623389 at /home/user/file.py (80ms)
(18524) Background analysis message: analyze
(18524) [BG(1)] analyzing: /home/user/file.py ...
(18524) [BG(1)]   checking: /home/user/file.py (24ms)
(18524) [BG(1)] analyzing: /home/user/file.py (24ms)
(18524) Background analysis message: resumeAnalysis
(18524) Background analysis message: getDiagnosticsForRange
(18524) Background analysis message: getDiagnosticsForRange
```

**Comment by erictraut** [comment]
The function `validation_not_none` in your example is called a "type guard function". If you want a static type checker to treat it as such, you need to use the `TypeGuard` type annotation introduced in [PEP 647](https://peps.python.org/pep-0647/).

Here's how that would look:

```python
from typing import List, Optional, TypeGuard, TypeVar

T = TypeVar("T")

def validation_not_none(field: Optional[T]) -> TypeGuard[T]:
    return field is not None
```

**Comment by vincentmasse** [comment]
I think TypeGuard is not available in python 3.8

**Comment by erictraut** [comment]
It's available in `typing_extensions`, which is a small library that back-ports type features to older versions of Python.
